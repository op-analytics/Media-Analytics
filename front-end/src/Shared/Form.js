import Button from '@material-ui/core/Button';
import FormControl from '@material-ui/core/FormControl';
import { makeStyles } from '@material-ui/core/styles';
import TextField from '@material-ui/core/TextField';
import PropTypes from 'prop-types';
import React, { useState } from 'react';

const useStyles = makeStyles(theme => ({
  form: {
    display: 'flex',
    flexDirection: 'row',
    flexWrap: 'wrap',
  },
  input: {
    border: 'solid 1px black',
  },
  formControl: {
    margin: theme.spacing(1),
  },
  button: {
    maxHeight: '40px',
    marginTop: '36px',
  },
}));

/**
 * Dynamic form
 * @component
 */
const Form = ({ formData, buttonText, onSubmit }) => {
  const [formInput, setFormInput] = useState({});
  const classes = useStyles();

  const handleChange = event => {
    const formInputCopy = { ...formInput };
    formInputCopy[event.target.name] = event.target.value;
    setFormInput(formInputCopy);
  };

  return (
    <form
      className={classes.form}
      onSubmit={e => {
        e.preventDefault();
        onSubmit(formInput);
      }}
    >
      {formData.map(f => (
        <FormControl key={f.name} className={classes.formControl}>
          <TextField
            className={classes.textField}
            margin="normal"
            label={`${f.label || f.name}:`}
            name={f.name}
            value={f.text}
            onChange={handleChange}
            placeholder={f.placeholder}
            required={f.required}
          />
        </FormControl>
      ))}
      <Button
        variant="outlined"
        color="primary"
        type="submit"
        className={classes.button}
      >
        {buttonText}
      </Button>
    </form>
  );
};

Form.defaultProps = { buttonText: 'Submit' };

Form.propTypes = {
  /**
   * The function to call on form submit
   * Will receive a data object consisting of all
   * formData names and there values
   */
  onSubmit: PropTypes.func.isRequired,
  /** The form data to build the form from */
  formData: PropTypes.arrayOf(
    PropTypes.shape({
      /** The fields name */
      name: PropTypes.string.isRequired,
      /** The fields label */
      label: PropTypes.string,
      /** The fields key */
      key: PropTypes.string,
      /** The fields placeholder */
      placeholder: PropTypes.string,
      /** Wether or not the field is required */
      required: PropTypes.bool,
    }),
  ).isRequired,
  /** The submit buttons text*/
  buttonText: PropTypes.string,
};

export default Form;
