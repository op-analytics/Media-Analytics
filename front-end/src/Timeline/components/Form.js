import React, { useState, useEffect } from "react";
import { makeStyles } from "@material-ui/core/styles";
import FormControl from "@material-ui/core/FormControl";
import FormHelperText from "@material-ui/core/FormHelperText";
import Input from "@material-ui/core/Input";
import InputLabel from "@material-ui/core/InputLabel";
import Slider from "@material-ui/core/Slider";

const useStyles = makeStyles(theme => ({
  formControl: {
    margin: theme.spacing(1)
  },
  slider: {
    paddingLeft: 2,

  }
}));

const marks = [{ value: 1960, label: "1960" }, { value: 2020, label: "2020" }];

export default function Form() {
  const [word, setWord] = useState("example");
  const [yearRange, setYearRange] = useState([1970, 2018]);
  const classes = useStyles();

  function WordInput() {
    return (
      <FormControl className={classes.formControl}>
        <InputLabel htmlFor="component-helper">Word</InputLabel>
        <Input
          id="component-helper"
          value={word}
          onChange={e => setWord(e.target.value)}
          aria-describedby="component-helper-text"
        />
        <FormHelperText id="component-helper-text">
          Can be a comma separated list
        </FormHelperText>
      </FormControl>
    );
  }

  return (
    <form >
      <WordInput />
      <div className={classes.slider}>
        <h4>Year range</h4>
        <Slider
          aria-label="airbnb slider"
          valueLabelDisplay="auto"
          value={yearRange}
          onChange={(_, newValue) => setYearRange(newValue)}
          getAriaValueText={value => `${value}`}
          min={1960}
          max={2020}
          marks={marks}
        />
      </div>
    </form>
  );
}
