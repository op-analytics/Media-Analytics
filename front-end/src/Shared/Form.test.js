import React from 'react';
import { shallow } from 'enzyme';
import Form from './Form';

const propsWith1Field = {
  onSubmit: jest.fn(),
  formData: [{ name: 'test' }],
};

const propsWith2Fields = {
  onSubmit: jest.fn(),
  formData: [{ name: 'test' }, { name: 'test2' }],
};

// eslint-disable-next-line react/jsx-props-no-spreading
const getWrapper = props => shallow(<Form {...props} />);

describe('Form', () => {
  it('should render a <form>', () => {
    expect(getWrapper(propsWith1Field).exists('form')).toBe(true);
  });

  it('renders correctly with one field', () => {
    // eslint-disable-next-line react/jsx-props-no-spreading
    const component = shallow(<Form {...propsWith1Field} />);
    expect(component).toMatchSnapshot();
  });

  it('renders correctly with more than one field', () => {
    // eslint-disable-next-line react/jsx-props-no-spreading
    const component = shallow(<Form {...propsWith2Fields} />);
    expect(component).toMatchSnapshot();
  });

  test('is fillable', () => {
    const testData = { yearFrom: '1980', word: 'man' };
    const wrapper = getWrapper(propsWith2Fields);

    const wordInput = wrapper.find('test');
    wordInput.value = testData.word;
    expect(wordInput.value).toBe('man');

    const yearFromInput = wrapper.find('test2');
    yearFromInput.value = testData.yearFrom;
    expect(yearFromInput.value).toBe('1980');
  });
});
