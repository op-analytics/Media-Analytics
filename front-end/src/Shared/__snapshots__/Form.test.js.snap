// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form renders correctly with more than one field 1`] = `
<form
  className="makeStyles-form-1"
  onSubmit={[Function]}
>
  <WithStyles(ForwardRef(FormControl))
    className="makeStyles-formControl-3"
    key="test"
  >
    <WithStyles(ForwardRef(TextField))
      label="test:"
      margin="normal"
      name="test"
      onChange={[Function]}
    />
  </WithStyles(ForwardRef(FormControl))>
  <WithStyles(ForwardRef(FormControl))
    className="makeStyles-formControl-3"
    key="test2"
  >
    <WithStyles(ForwardRef(TextField))
      label="test2:"
      margin="normal"
      name="test2"
      onChange={[Function]}
    />
  </WithStyles(ForwardRef(FormControl))>
  <WithStyles(ForwardRef(Button))
    className="makeStyles-button-4"
    color="primary"
    type="submit"
    variant="outlined"
  >
    Submit
  </WithStyles(ForwardRef(Button))>
</form>
`;

exports[`Form renders correctly with one field 1`] = `
<form
  className="makeStyles-form-1"
  onSubmit={[Function]}
>
  <WithStyles(ForwardRef(FormControl))
    className="makeStyles-formControl-3"
    key="test"
  >
    <WithStyles(ForwardRef(TextField))
      label="test:"
      margin="normal"
      name="test"
      onChange={[Function]}
    />
  </WithStyles(ForwardRef(FormControl))>
  <WithStyles(ForwardRef(Button))
    className="makeStyles-button-4"
    color="primary"
    type="submit"
    variant="outlined"
  >
    Submit
  </WithStyles(ForwardRef(Button))>
</form>
`;
